### ISSUE TITLE
942510 false positive

### ISSUE THREAD

**Issue by aramhovsepyan** [SpiderLabs/owasp-modsecurity-crs#1524]
## Type of Issue
It seems that the rule 942510 is blocking a trivial 'test' input.

## Your Environment

* CRS version (e.g. v3.0.2): 3.2/dev
* ModSecurity version (e.g. 2.9.2): 2.9.3
* Web Server and version (e.g. apache 2.4.27): 2.4.29
* Operating System and version: ubuntu

## Audit log in JSON
{"transaction":{"time":"27/Aug/2019:11:43:00 +0200","transaction_id":"XWT7JM0kvDID-KM-YeLu6gAAAAA","remote_address":"192.168.93.1","remote_port":60765,"local_address":"192.168.93.130","local_port":80},"request":{"request_line":"POST /admin/videoevaluation/modify/id/1 HTTP/1.1","headers":{"Host":"aramvm.local","Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8","Accept-Encoding":"gzip, deflate","Accept-Language":"en-us","Content-Type":"application/x-www-form-urlencoded","Origin":"http://aramvm.local","User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.2 Safari/605.1.15","Connection":"keep-alive","Upgrade-Insecure-Requests":"1","Referer":"http://aramvm.local/admin/videoevaluation/modify/id/1","Content-Length":"134","body":["csrf=b11fde04446d1fbac834d16b362ce8df-ca3e303938f85f5a5c6af3735baf6101&title=test&type=0&intro=this+is+a+%27test%27+input&submit=&id=1"]},"response":{"protocol":"HTTP/1.1","status":403,"headers":{"Content-Length":"84","Keep-Alive":"timeout=5, max=100","Connection":"Keep-Alive","Content-Type":"text/html; charset=iso-8859-1"}},"audit_data":{"messages":["Access denied with code 403 (phase 2). [file \"/etc/modsecurity/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf\"] [line \"555\"] [id \"942510\"] [msg \"SQLi bypass attempt by ticks or backticks detected.\"] [data \"Matched Data: 'test' found within ARGS:intro: this is a 'test' input\"] [severity \"CRITICAL\"] [ver \"OWASP_CRS/3.2.0\"] [tag \"application-multi\"] [tag \"language-multi\"] [tag \"platform-multi\"] [tag \"attack-sqli\"] [tag \"OWASP_CRS\"] [tag \"OWASP_CRS/WEB_ATTACK/SQL_INJECTION\"] [tag \"WASCTC/WASC-19\"] [tag \"OWASP_TOP_10/A1\"] [tag \"OWASP_AppSensor/CIE1\"] [tag \"PCI/6.5.2\"]"],"action":{"intercepted":true,"phase":2,"message":"Pattern match \"(?:(['`])((?:[\\\\w\\\\s=_\\\\-+{}()<@]){2,29}|(?:[A-Za-z0-9+\\\\/]{4})+(?:[A-Za-z0-9+\\\\/]{2}==|[A-Za-z0-9+\\\\/]{3}=)?)\\\\1)\" at ARGS:intro."},"stopwatch":{"p1":4210,"p2":34815,"p3":0,"p4":0,"p5":578,"sr":143,"sw":105,"l":0,"gc":0},"producer":["ModSecurity for Apache/2.9.2 (http://www.modsecurity.org/)","OWASP_CRS/3.1.0"],"server":"Apache/2.4.29 (Ubuntu)","engine_mode":"ENABLED"}}

## Confirmation
[X] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by dune73** [comment]
Confirmed. This is a false positive. Thank you for reporting.

```
$>cat /tmp/tmp2
intro=this is a 'test' input
$>send-payload-pls.sh /tmp/tmp2
Sending the following payload at multiple paranoia levels: /tmp/tmp2

--- Paranoia Level 1 ---

942510 SQLi bypass attempt by ticks or backticks detected.

Total Incoming Score: 5

--- Paranoia Level 2 ---

942510 SQLi bypass attempt by ticks or backticks detected.

Total Incoming Score: 5

--- Paranoia Level 3 ---

920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
942510 SQLi bypass attempt by ticks or backticks detected.

Total Incoming Score: 15

--- Paranoia Level 4 ---

920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
920273 PL4 Invalid character in request (outside of very strict set)
920273 PL4 Invalid character in request (outside of very strict set)
942432 PL4 Restricted SQL Character Anomaly Detection (args): # of special characters exceeded (2)
942510 SQLi bypass attempt by ticks or backticks detected.

Total Incoming Score: 28

```

**Comment by spartantri** [comment]
@dune73 I like your script, you should consider adding it to crs util scripts :)

**Comment by dune73** [comment]
Thank you. Yes, probably. But it's a bit of a hack so far - and it depends on a very specific configuration of the webserver - that you do not want in prod :):
```
#!/bin/bash
#
# Script to post a payload against a local webserver at each paranoia level
#
# Note: Webserver has to be prepared to take desired PL as Request Header "PL"
#

if [ ! -z "$1" ]; then
        PAYLOAD="$1"
else
        echo "Please submit payload as parameter. This is fatal. Aborting."
        exit 1
fi

echo "Sending the following payload at multiple paranoia levels: $PAYLOAD"
echo

for PL in 1 2 3 4; do 
        echo "--- Paranoia Level $PL ---" 
        echo
        if [ -f "$PAYLOAD" ]; then
                curl localhost:40080 --data-binary "@$PAYLOAD" -H "PL: $PL" -o /dev/null -s
        else
                curl localhost:40080 -d "$PAYLOAD" -H "PL: $PL" -o /dev/null -s
        fi
        grep $(tail -1 /apache/logs/access.log | cut -d\" -f11 | cut -b2-26) /apache/logs/error.log | sed -e "s/.*\[id \"//" -e "s/\(......\).*\[msg \"/\1 /" -e "s/\"\].*//" -e "s/(Total .*/(Total ...) .../" -e "s/Incoming and Outgoing Score: [0-9]* [0-9]*/Incoming and Outgoing Score: .../" | sed -e "s/901001\|901450\|905100\|905110\|910000\|910100\|910150\|910160\|910170\|910180\|911100\|912120\|912170\|913100\|913110\|913120\|920100\|920120\|920130\|920140\|920160\|920170\|920180\|920190\|920210\|920220\|920240\|920250\|920260\|920270\|920280\|920290\|920310\|920311\|920330\|920340\|920350\|920360\|920370\|920380\|920390\|920400\|920410\|920420\|920430\|920440\|920450\|921100\|921110\|921120\|921130\|921140\|921150\|921160\|930100\|930110\|930120\|930130\|931100\|931110\|931120\|932100\|932105\|932110\|932115\|932120\|932130\|932140\|932150\|932160\|932170\|932171\|933100\|933110\|933120\|933130\|933140\|933150\|933160\|933170\|933180\|941100\|941110\|941120\|941130\|941140\|941150\|941160\|941170\|941180\|941190\|941200\|941210\|941220\|941230\|941240\|941250\|941260\|941270\|941280\|941290\|941300\|941310\|941350\|942100\|942140\|942160\|942170\|942190\|942220\|942230\|942240\|942250\|942270\|942280\|942290\|942320\|942350\|942360\|943100\|943110\|943120\|944100\|944110\|944120\|944130\|949100\|949110\|950130\|951110\|951120\|951130\|951140\|951150\|951160\|951170\|951180\|951190\|951200\|951210\|951220\|951230\|951240\|951250\|951260\|952100\|952110\|953100\|953110\|953120\|954100\|954110\|954120\|954130\|959100\|980100\|980110\|980120\|980130\|980140\|9001000\|9001110\|9001112\|9001114\|9001116\|9001120\|9001122\|9001124\|9001126\|9001128\|9001140\|9001150\|9001170\|9001180\|9001182\|9001184\|9001200\|9001202\|9001204\|9001206\|9001208\|9001210\|9001212\|9001214\|9001216\|9002000\|9002001\|9002100\|9002120\|9002130\|9002150\|9002160\|9002200\|9002400\|9002401\|9002410\|9002420\|9002520\|9002530\|9002540\|9002700\|9002710\|9002720\|9002730\|9002740\|9002750\|9002800\|9002810\|9002820\|9002900/& PL1/" -e "s/912171\|913101\|913102\|920200\|920201\|920230\|920271\|920300\|920320\|921151\|931130\|933151\|941320\|941330\|941340\|942110\|942120\|942130\|942150\|942180\|942200\|942210\|942260\|942300\|942310\|942330\|942340\|942370\|942380\|942390\|942400\|942410\|942430\|942440\|942450\|944200\|944210\|944240\|944250\|950100/& PL2/" -e "s/920272\|921180\|933111\|933131\|933161\|942251\|942420\|942431\|942460\|944300/& PL3/" -e "s/920202\|920273\|920274\|920460\|942421\|942432/& PL4/" | sort -k2
        echo
        echo -n "Total Incoming Score: "
        tail -1 /apache/logs/access.log | cut -d\" -f11 | cut -d\  -f14 | tr "-" "0" 
        echo
done

```

**Comment by lifeforms** [comment]
Thanks @aramhovsepyan for the false positive report. Your report arrived just in time for our fixing it before the v3.2 release next month. A fix is in #1528.
